scripts/doc-checker.py

```diff
@@ -36,55 +48,83 @@ def validate_object(import_path):
-    methods = []
-    for _class in classes:
-        for node in _class.body:
-            if is_public_func(node):
-                methods.append(f"{module_name}.{_class.name}.{node.name}")
-
-    print(f"NUMPYDOC OUTPUT FOR {path} - CAN BE EMPTY")
+def numpydoc_validate(path: pathlib.Path) -> int:
+    """
+    Perform numpydoc checks.
+
+    Parameters
+    ----------
+    path : pathlib.Path
+
+    Returns
+    -------
+    exit_status : int
+    """
     exit_status = 0
-    # numpydoc docstrings validation
-    # docstrings are taken dynamically
-    to_validate = (
-        [module_name]
-        + [f"{module_name}.{x.name}" for x in (functions + classes)]
-        + methods
-    )
-    if any(list(map(validate_object, to_validate))):
-        exit_status = 1
+
+    if path.is_file():
+        walker = ((str(path.parent), [], [path.name]),)
+    else:
+        walker = os.walk(path)
+
+    for root, _, files in walker:
+        if "__pycache__" in root:
+            continue
+        for f in files:
+            if not f.endswith(".py"):
+                continue
+
+            current_path = os.path.join(root, f)
+            # get importable name
+            module_name = current_path.replace("/", ".").replace("\\", ".")
+            module_name = module_name.replace(".py", "")
+
+            with open(current_path) as fd:
+                file_contents = fd.read()
+
+            # using static parsing for collecting module, functions, classes and its methods
+            module = ast.parse(file_contents)
+
+            def is_public_func(node):
+                return isinstance(node, ast.FunctionDef) and (
+                    not node.name.startswith("__") or node.name.endswith("__")
+                )
+
+            functions = [node for node in module.body if is_public_func(node)]
+            classes = [node for node in module.body if isinstance(node, ast.ClassDef)]
+            methods = []
+            for _class in classes:
+                for node in _class.body:
+                    if is_public_func(node):
+                        methods.append(f"{module_name}.{_class.name}.{node.name}")
+
+            print(f"NUMPYDOC OUTPUT FOR {current_path} - CAN BE EMPTY")
+            # numpydoc docstrings validation
+            # docstrings are taken dynamically
+            to_validate = (
+                [module_name]
+                + [f"{module_name}.{x.name}" for x in (functions + classes)]
+                + methods
+            )
+            if any(list(map(validate_object, to_validate))):
+                exit_status = 1
     return exit_status
 
 
-def pydocstyle_validate(path, add_ignore):
+def pydocstyle_validate(path: pathlib.Path, add_ignore: List[str]) -> int:
+    """
+    Perform pydocstyle checks.
+
+    Parameters
+    ----------
+    path : pathlib.Path
+    add_ignore : List[int]
+
+    Returns
+    -------
+    int
+        Return code of pydocstyle subprocess.
+    """
     result = subprocess.run(
```

Comment:
isn't there a better way of executing `pydocstyle` by importing it and using some API?..

Reply:
This module contains a public api for checking only the list of files, not directories. Checking directories is available only when the script is explicitly called as the main one.

Bottom line, this can be done using imports from the module, but it will be a little more complicated and more room for errors. If you don't mind, then I would not like to rewrite.

Reply:
Well, you're already validating using an `os.walk()` method... but we can improve later, it's certainly not a blocker.